options(stringsAsFactors=FALSE)
library(optparse)
library(ggplot2)
library(gplots)
library(reshape2)
theme_set(theme_bw(base_size = 12))


option_list <- list(
  make_option(
    c("--kin0"), type = "character", help = "Between-family kinship table generated by running king in PLINK."
  ),
  make_option(
    c("--kinMat"), type = "character", help = "Between-family kinship matrix generated by running king in PLINK."
  ),
  make_option(
    c("--kinIds"), type = "character", help = "Between-family kinship matrix subjet IDs generated by running king in PLINK."
  ),
  make_option(
    c("--genoraw"), type = "character", help = "Plink-generated (using --recodeA) alt allele count matrix."
  ),
  make_option(
    c("--donors"), type = "character", help = "Genotyped donors."
  ),
  make_option(
    c("--outdir"), type = "character", help = "Output destination for results."
  )
)

option_parser <- OptionParser(usage = "usage: Rscript %prog [options]",
                              option_list = option_list, add_help_option = TRUE)
opts <- parse_args(option_parser)

### Testing
# opts = list()
# opts$kin0 =  "results/multiome/demultiplex_gex_auto/4/souporcell_clusters_and_donors_filtered.kin0"
# opts$kinMat = "results/multiome/demultiplex_gex_auto/4/souporcell_clusters_and_donors_filtered.king"
# opts$kinIds = "results/multiome/demultiplex_gex_auto/4/souporcell_clusters_and_donors_filtered.king.id"
# opts$genoraw = "results/multiome/demultiplex_gex_auto/4/souporcell_clusters_and_donors_filtered.raw"
# opts$outdir = "results/multiome/demultiplex_gex_auto/4"
# opts$donors = "HPAP105,HPAP093,ICRH139,ICRH142,HPAP107"
#


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Using Kinship coefficients
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

### Read in as data frame
df = read.table(opts$kin0, header=FALSE)
names(df) <- c("FID1","ID1","FID2","ID2","N_SNP","HetHet","IBS0","Kinship")
df$Kinship_bounded = df$Kinship
df$Kinship_bounded[!is.finite(df$Kinship)] <- 0
df$Kinship_bounded[df$Kinship < 0] <- 0


### Create other half of triangle
df2 = df
df2$ID1 = df$ID2
df2$FID1 = df$FID2
df2$ID2 = df$ID1
df2$FID2 = df$FID1

### Create square
sq = rbind(df, df2)

### Create categorical kinship
sq$Kinship_cat = NA
sq$Kinship_cat[sq$Kinship_bounded > 0.35] <- "MZ"
sq$Kinship_cat[sq$Kinship_bounded > 0.177 & sq$Kinship_bounded < 0.35] <- "1D"
sq$Kinship_cat[sq$Kinship_bounded > 0.088 & sq$Kinship_bounded < 0.177] <- "2D"
sq$Kinship_cat[sq$Kinship_bounded < 0.088] <- "Unrelated"
sq$Kinship_cat = factor(sq$Kinship_cat, levels = c("Unrelated","2D", "1D","MZ"))

### Create batch indicator
souporcell_ids = unique(sq$ID1[grep("souporcell",sq$ID1)])
donor_ids = unique(sq$ID1[!grepl("souporcell",sq$ID1)])
## !!!!!!!!!!!! UPDATE THIS
batch = gsub("-hg38","", gsub("5124-", "", sapply(sapply(souporcell_ids, strsplit, split="_"), `[`, 2)))
#batch = sapply(X=souporcell_ids, FUN=function(x) {gsub(pattern="_souporcell_.*$", replace="", x, perl=TRUE)})
design = data.frame(ID = c(souporcell_ids, donor_ids), batch=c(batch, rep("donor",length(donor_ids))))
row.names(design) = design$ID
sq$Batch1 = design[sq$ID1,"batch"]
sq$Batch2 = design[sq$ID2,"batch"]

donors = unlist(strsplit(opts$donors, split=","))
#sq_sub = sq[sq$ID2 %in% donors & !sq$ID1 %in% donor_ids,]
wrong_donors = donor_ids[!donor_ids %in% donors]

sq_sub = sq[sq$ID2 %in% donor_ids & !sq$ID1 %in% donor_ids,]
sq_sub$donor_right_wrong = NA
sq_sub$donor_right_wrong[sq_sub$ID2 %in% donors] <- "Right"
sq_sub$donor_right_wrong[sq_sub$ID2 %in% wrong_donors] <- "Wrong"
sq_sub$donor_right_wrong = factor(sq_sub$donor_right_wrong, levels = c("Right", "Wrong"))

#set HPAP107 to missing for batches 1-4
sq_sub$Kinship_bounded[sq_sub$ID2 == "HPAP107" & sq_sub$Batch1 %in% c("NM-1","NM-2","NM-3","NM-4")] <- NA
sq_sub$Kinship_cat[sq_sub$ID2 == "HPAP107" & sq_sub$Batch1 %in% c("NM-1","NM-2","NM-3","NM-4")] <- NA


# png(file.path(opts$outdir,"corrplot_kinship.png"), width=1200, height=900)
# ggplot(sq) +
#   geom_tile(aes(x=ID1, y=ID2, fill=Kinship_bounded)) +
#   theme(axis.text.x = element_text(angle=90)) +
#    scale_fill_continuous(limits = c(0, 0.5))
# dev.off()
#
# png(file.path(opts$outdir,"corrplot_kinship_factor.png"), width=1200, height=900)
# ggplot(sq) +
#   geom_tile(aes(x=ID1, y=ID2, fill=Kinship_cat)) +
#   theme(axis.text.x = element_text(angle=90))
# dev.off()


if (length(unique(batch))>1) {


  png(file.path(opts$outdir,"corrplot_kinship_by_donor.png"), width=1200, height=500)
  p1 = ggplot(sq_sub) +
    geom_tile(aes(x=ID1, y=ID2, fill=Kinship_bounded)) +
    facet_grid(donor_right_wrong~Batch1, scales="free") +
    theme(axis.text.x = element_text(angle=90)) +
    scale_fill_continuous(limits = c(0, 0.5))
  print(p1)
  dev.off()

  png(file.path(opts$outdir,"corrplot_kinship_factor_by_donor.png"), width=1200, height=500)
  p2 = ggplot(sq_sub) +
    geom_tile(aes(x=ID1, y=ID2, fill=Kinship_cat)) +
    facet_grid(donor_right_wrong~Batch1, scales="free") +
    theme(axis.text.x = element_text(angle=90))
  print(p2)
  dev.off()


} else {


  png(file.path(opts$outdir,"corrplot_kinship_by_donor.png"), width=1200, height=500)
  p1 = ggplot(sq_sub) +
    geom_tile(aes(x=ID1, y=ID2, fill=Kinship_bounded)) +
    theme(axis.text.x = element_text(angle=90)) +
    scale_fill_continuous(limits = c(0, 0.5))
  print(p1)
  dev.off()

  png(file.path(opts$outdir,"corrplot_kinship_factor_by_donor.png"), width=1200, height=500)
  p2 = ggplot(sq_sub) +
    geom_tile(aes(x=ID1, y=ID2, fill=Kinship_cat)) +
    theme(axis.text.x = element_text(angle=90))
  print(p2)
  dev.off()


}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Using allele count correlation matrix
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rm(list=ls()[!ls() %in% "opts"])

library(data.table)
genoDF = fread(opts$genoraw, header=TRUE)
genoMat = as.matrix(genoDF[, 7:ncol(genoDF)])
row.names(genoMat) = genoDF$IID

genoMatTP = t(genoMat)
genoCorrMat = cor(genoMatTP, use="pairwise.complete.obs", method="spearman")

pc = reshape2::melt(genoCorrMat)
names(pc) <- c("ID1", "ID2", "r")
pc$ID1 = as.character(pc$ID1)
pc$ID2 = as.character(pc$ID2)

pc$r_factor = NA
pc$r_factor[pc$r < 0.2] <- "   < 0.2"
pc$r_factor[pc$r >= 0.2 & pc$r < 0.5] <- "0.2 - 0.5"
pc$r_factor[pc$r >= 0.5 & pc$r <= 0.9] <- "0.5 - 0.9"
pc$r_factor[pc$r > 0.9] <- "> 0.9"
pc$r_factor = factor(pc$r_factor, levels = c("   < 0.2", "0.2 - 0.5", "0.5 - 0.9", "> 0.9"))

pc$r_bounded = pc$r
pc$r_bounded[pc$r < 0] <- 0

souporcell_ids = unique(pc$ID1[grep("souporcell",pc$ID1)])
donor_ids = unique(pc$ID1[!grepl("souporcell",pc$ID1)])
## !!!!!!!!!!!! UPDATE THIS
batch = gsub("-hg38","", gsub("5124-", "", sapply(sapply(souporcell_ids, strsplit, split="_"), `[`, 2)))
#batch = sapply(X=souporcell_ids, FUN=function(x) {gsub(pattern="_souporcell_.*$", replace="", x, perl=TRUE)})
design = data.frame(ID = c(souporcell_ids, donor_ids), batch=c(batch, rep("donor",length(donor_ids))))
row.names(design) = design$ID
pc$Batch1 = design[pc$ID1,"batch"]
pc$Batch2 = design[pc$ID2,"batch"]


donors = unlist(strsplit(opts$donors, split=","))
wrong_donors = donor_ids[!donor_ids %in% donors]

#pc_sub_wrong = pc[pc$ID2 %in% donors & !pc$ID1 %in% donor_ids,]
pc_sub = pc[pc$ID2 %in% donor_ids & !pc$ID1 %in% donor_ids,]
pc_sub$donor_right_wrong = NA
pc_sub$donor_right_wrong[pc_sub$ID2 %in% donors] <- "Right"
pc_sub$donor_right_wrong[pc_sub$ID2 %in% wrong_donors] <- "Wrong"
pc_sub$donor_right_wrong = factor(pc_sub$donor_right_wrong, levels = c("Right", "Wrong"))

#set HPAP107 to missing for batches 1-4
pc_sub$r_bounded[pc_sub$ID2 == "HPAP107" & pc_sub$Batch1 %in% c("NM-1","NM-2","NM-3","NM-4")] <- NA
pc_sub$r_factor[pc_sub$ID2 == "HPAP107" & pc_sub$Batch1 %in% c("NM-1","NM-2","NM-3","NM-4")] <- NA

# png(file.path(opts$outdir,"corrplot_r.png"), width=1200, height=800)
# ggplot(pc) +
#   geom_tile(aes(x=ID1, y=ID2, fill=r_bounded)) +
#   theme(axis.text.x = element_text(angle=90)) +
#   scale_fill_continuous(limits = c(0, 1))
# dev.off()
#
# png(file.path(opts$outdir,"corrplot_r_factor.png"), width=1200, height=800)
# ggplot(pc) +
#   geom_tile(aes(x=ID1, y=ID2, fill=r_factor)) +
#   theme(axis.text.x = element_text(angle=90))
# dev.off()




if (length(unique(batch))>1) {

  png(file.path(opts$outdir,"corrplot_r_by_donor.png"), width=1200, height=500)
  p1 = ggplot(pc_sub) +
    geom_tile(aes(x=ID1, y=ID2, fill=r_bounded)) +
    facet_grid(donor_right_wrong~Batch1, scales="free") +
    theme(axis.text.x = element_text(angle=90)) +
    scale_fill_continuous(limits = c(0, 1))
  print(p1)
  dev.off()

  png(file.path(opts$outdir,"corrplot_r_factor_by_donor.png"), width=1200, height=500)
  p2 = ggplot(pc_sub) +
    geom_tile(aes(x=ID1, y=ID2, fill=r_factor)) +
    facet_grid(donor_right_wrong~Batch1, scales="free") +
    theme(axis.text.x = element_text(angle=90))
  print(p2)
  dev.off()

} else {

  png(file.path(opts$outdir,"corrplot_r_by_donor.png"), width=1200, height=500)
  p1 = ggplot(pc_sub) +
    geom_tile(aes(x=ID1, y=ID2, fill=r_bounded)) +
    theme(axis.text.x = element_text(angle=90)) +
    scale_fill_continuous(limits = c(0, 1))
  print(p1)
  dev.off()


  png(file.path(opts$outdir,"corrplot_r_factor_by_donor.png"), width=1200, height=500)
  p2 = ggplot(pc_sub) +
    geom_tile(aes(x=ID1, y=ID2, fill=r_factor)) +
    theme(axis.text.x = element_text(angle=90))
  print(p2)
  dev.off()


}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Incorporate information on number of nuclei per cluster
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#clusters = read.table("results/multiome/souporcell_gex_auto/Sample_5124-NM-5-hg38/4/clusters.tsv", header=TRUE)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Heatmap/cluster samples based on relatedness metrics
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Wasn't exactly sure what the heatmap function was doing, so made one manually
# following this example:
#http://www.sthda.com/english/wiki/ggplot2-quick-correlation-matrix-heatmap-r-software-and-data-visualization
# reorder_cormat <- function(cormat){
#   dd <- as.dist((1-cormat)/2)
#   hc <- hclust(dd)
#   return(cormat[hc$order, hc$order])
# }
#
# makeCorMatHeatmap = function(x) {
#   df = reshape2::melt(reorder_cormat(x))
#   p = ggplot(df, aes(x=Var1, y=Var2, fill=value)) +
#     geom_tile(color = "white")+
#     scale_fill_gradient2(low = "white", high = "red", limit = c(0,1), space = "Lab", name="Genotype\nCorrelation") +
#     theme(axis.text.x = element_text(angle = 90, vjust = 1, size = 12, hjust = 1))+
#     coord_fixed()
#   print(p)
# }
#
# makeKinMatHeatmap = function(x) {
#   df = reshape2::melt(reorder_cormat(x))
#   p = ggplot(df, aes(x=Var1, y=Var2, fill=value)) +
#     geom_tile(color = "white")+
#     scale_fill_gradient2(low = "white", high = "red", limit = c(0,0.5), space = "Lab", name="Kinship Coefficient") +
#     theme(axis.text.x = element_text(angle = 90, vjust = 1, size = 12, hjust = 1))+
#     coord_fixed()
#   print(p)
# }
#
#
# ### Using kinship coefficients
# kinMat = as.matrix(read.table(opts$kinMat, header=FALSE))
# ids = read.table(opts$kinIds, header=TRUE, comment.char="~")$IID
# row.names(kinMat) <- ids 
# colnames(kinMat) <- ids
# kinMat[!is.finite(kinMat)] <- 0
# kinMat[kinMat<0] <- 0
#
# png(file.path(opts$outdir,"corrplot_kinship_heatmap.png"))
# #heatmap(kinMat)
# makeKinMatHeatmap(kinMat)
# #makeKinMatHeatmap(kinMat[c(donors,souporcell_ids),c(donors,souporcell_ids)])
# dev.off()
#
#
# ### Using genotype correlations
# genoMatTP = t(genoMat)
# genoCorrMat = cor(genoMatTP, use="pairwise.complete.obs")
# genoCorrMat[genoCorrMat<0] <- 0
# genoCorrMat[is.na(genoCorrMat)] <-0
#
# png(file.path(opts$outdir,"corrplot_r_heatmap.png"))
# makeCorMatHeatmap(genoCorrMat)
# #makeCorMatHeatmap(genoCorrMat[c(donors,souporcell_ids),c(donors,souporcell_ids)])
# dev.off()
#
#



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Plot cell allele support PCA (trying to replicate figures in paper)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
